src/async/imap/MCIMAPAsyncSession.cpp

```diff
@@ -306,58 +317,63 @@ IMAPAsyncConnection * IMAPAsyncSession::sessionForFolder(String * folder, bool u
 
 IMAPAsyncConnection * IMAPAsyncSession::availableSession()
 {
-    if (mMaximumConnections == 0) {
-        for(unsigned int i = 0 ; i < mSessions->count() ; i ++) {
-            IMAPAsyncConnection * s = (IMAPAsyncConnection *) mSessions->objectAtIndex(i);
-            if (s->operationsCount() == 0)
-                return s;
-        }
-        IMAPAsyncConnection * chosenSession = session();
-        mSessions->addObject(chosenSession);
+    // try find existant session with empty queue for reusing.
+    IMAPAsyncConnection * chosenSession = sessionWithMinQueue(false, NULL);
+    if ((chosenSession != NULL) && (chosenSession->operationsCount() == 0)) {
         return chosenSession;
     }
-    else {
-        IMAPAsyncConnection * chosenSession = NULL;
-        unsigned int minOperationsCount = 0;
-        for(unsigned int i = 0 ; i < mSessions->count() ; i ++) {
-            IMAPAsyncConnection * s = (IMAPAsyncConnection *) mSessions->objectAtIndex(i);
-            if (chosenSession == NULL) {
-                chosenSession = s;
-                minOperationsCount = s->operationsCount();
-            }
-            else if (s->operationsCount() < minOperationsCount) {
-                chosenSession = s;
-                minOperationsCount = s->operationsCount();
-            }
-        }
-        if (mSessions->count() < mMaximumConnections) {
-            if ((chosenSession != NULL) && (minOperationsCount == 0)) {
-                return chosenSession;
-            }
-            chosenSession = session();
-            mSessions->addObject(chosenSession);
-            return chosenSession;
-        }
-        else {
-            return chosenSession;
-        }
+
+    // create new session, if maximum connections limit does not reached yet.
+    if ((mMaximumConnections == 0) || (mSessions->count() < mMaximumConnections)) {
+        chosenSession = session();
+        mSessions->addObject(chosenSession);
+        return chosenSession;
     }
+
+    // otherwise returns existant session with minimum size of queue.
+    return chosenSession;
 }
 
 IMAPAsyncConnection * IMAPAsyncSession::matchingSessionForFolder(String * folder)
 {
-    for(unsigned int i = 0 ; i < mSessions->count() ; i ++) {
-        IMAPAsyncConnection * currentSession = (IMAPAsyncConnection *) mSessions->objectAtIndex(i);
-        if (currentSession->lastFolder() != NULL) {
-            if (currentSession->lastFolder()->isEqual(folder)) {
-                return currentSession;
+    // try find session with minimum size of queue among selected to the folder.
+    IMAPAsyncConnection * chosenSelectedSession = sessionWithMinQueue(true, folder);
+    if (chosenSelectedSession != NULL) {
+        return chosenSelectedSession;
+    }
+
+    // try find session with minimum size of queue among non-selected to any folder ones.
+    IMAPAsyncConnection * chosenNonSelectedSession = sessionWithMinQueue(true, NULL);
```

Comment:
The main idea is in reusing of available session already selected to needed folder.
Since the SELECT command is very difficult for the server and can take a very long time, we have to minimize the number of current folder changes.

In case of folder == NULL in sessionForFolder() is used another branch of code:
    ```
if (folder == NULL) {
        return availableSession();
    }
```